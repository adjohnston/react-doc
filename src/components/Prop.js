import React from 'react'
import PropTypes from 'prop-types'
import classString from 'src/helpers/class-string'
import Input from 'src/components/Input'

const Prop = ({name, defs = {}, ...restProps}) => {
  const {
    type,
    required,
    description
  } = defs

  let typeElement
  if (type) {
    typeElement = (
      <span
        className={classString('__prop-type')}>
        ({type})
      </span>
    )
  } else {
    typeElement = (
      <span
        className={classString('__prop-type-missing')}>
        Missing type
      </span>
    )
  }

  let requiredElement
  if (required) {
    requiredElement = (
      <span
        className={classString('__prop-required')}>
        Required
      </span>
    )
  }

  let descriptionElement
  if (description) {
    descriptionElement = (
      <div>
        {description}
      </div>
    )
  }

  let inputElement
  if (type === 'string' || type === 'number' || type === 'bool') {
    inputElement = (
      <Input
        type={type}
        name={name}
        {...restProps} />
    )
  }

  return (
    <li>
      <div>
        {name}
      </div>

      <div>
        {typeElement}

        {requiredElement}
      </div>

      {descriptionElement}
      {inputElement}
    </li>
  )
}

Prop.propTypes = {
  name: PropTypes.string.isRequired,
  defs: PropTypes.shape({
    type: PropTypes.string.isRequired,
    required: PropTypes.bool,
    defaultValue: PropTypes.node,
    description: PropTypes.string
  }).isRequired,
  restProps: PropTypes.object
}

export default Prop
