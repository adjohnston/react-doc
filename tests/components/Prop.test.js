import React from 'react'
import renderer from 'react-test-renderer'
import Prop from 'components/Prop'

test('renders correctly', () => {
  const mockProps = {
    name: 'className',
    defs: {
      type: 'string',
      required: true,
      defaultValue: 'button',
      description: 'Add a class to the element'
    },
    value: 'button',
    onChange: () => {}
  }

  const tree = renderer
    .create(<Prop {...mockProps} />)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

test('renders a toggle element with a type of bool', () => {
  const mockProps = {
    name: 'isActive',
    defs: {
      type: 'bool',
      required: true,
      defaultValue: false,
      description: 'isActive indicates whether the component is active or not'
    },
    value: true,
    onChange: () => {}
  }

  const tree = renderer
    .create(<Prop {...mockProps} />)
    .toJSON()

  expect(tree).toMatchSnapshot()
})
