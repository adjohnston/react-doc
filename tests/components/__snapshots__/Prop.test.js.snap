// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<li>
  <div>
    className
  </div>
  <div>
    <b>
      string
    </b>
    <i>
       - is required
    </i>
  </div>
  <input
    className="stncl__input"
    onChange={[Function]}
    type="text"
    value="button"
  />
</li>
`;
